File: python/symmetric-binary-tree.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

class Tree:
    def __init__(self, root):
        self.root = root

class Solution:
    def isSymmetric(self, root):
        if root is None:
            return True
        return self.isMirror(root.left, root.right)
    
    def isMirror(self, root1, root2):
        if root1 is None and root2 is None:
            return True
        if root1 is not None and root2 is not None:
            return root1.val == root2.val and self.isMirror(root1.left, root2.right) and self.isMirror(root1.right, root2.left)
        return False
    
root = Node(1)
root.left = Node(2)
root.left.left = Node(3)
root.left.right = Node(4)
root.right = Node(2)
root.right.left = Node(4)
root.right.right = Node(3)

print(Solution().isSymmetric(root))
    

File: javascript/symmetric-binary-tree.js
'use strict';

const isSymmetric = (root) => {
    const isMirror = (p, q) => {
        if (!p && !q) {
            return true;
        }
        
        if (!p || !q) {
            return false;
        }
        
        return p.val === q.val && isMirror(p.left, q.right) && isMirror(p.right, q.left);
    };
    
    return isMirror(root, root);
};

const root = {
    val: 1,
    left: {
        val: 2,
        left: {
        val: 3,
        left: null,
        right: null,
        },
        right: null,
    },
    right: {
        val: 2,
        left: null,
        right: {
        val: 3,
        left: null,
        right: null,
        },
    },
};

console.log(isSymmetric(root)); // true